### ISSUE TITLE
rsync over websocket sometimes fails

### ISSUE THREAD

**Issue by tych0** [lxc/lxd#2494]
it looks like the real `nc` isn't much better but we shouldn't do this. I believe:

https://github.com/tych0/lxd/commit/033e9627fc8fc0d0fb922757e9999b12ebc9a928

will fix it, but I don't have a reproducer. We have someone who does, this is just a reminder to track it when he gets back.

**Comment by stgraber** [comment]
Did we hear any other report of this? has any progress been made on this issue?

**Comment by tych0** [comment]
Yeah, I think we've heard other people say stuff like this. I can't reproduce it though.

**Comment by ChrisLundquist** [comment]
I *may* be hitting this. I was trying to live migrate a container with criu, but it didn't get that far.
The observed behavior was it started rsyncing some stuff, then died.
```
 lxc move --verbose --debug nuc:es-1 es-1
# ...
 error: websocket: bad handshake
```
Debug output.
https://gist.github.com/ChrisLundquist/2850dba12397e78648e638f5cabfc073

One system is ext4 backed, the other is btrfs. Both have the same version:
```
clundquist@NUCALASCAGE:~$ lxd --version
2.0.8
clundquist@NUCALASCAGE:~$ uname -a
Linux NUCALASCAGE 4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

```
clundquist@Draupnir:~$ lxd --version
2.0.8
clundquist@Draupnir:~$ uname -a
Linux Draupnir 4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by stgraber** [comment]
Look at /var/log/lxd/lxd.log on both hosts, I bet you'll see a CRIU error in there. I'm not sure why this leads to that "bad handshake" instead of the error being surfaced back to the client though...

@tych0

**Comment by ChrisLundquist** [comment]
`/var/log/lxd/lxd.log` is pretty clean, just mentions the migration failed, but log level is only at info.

Moving *from* this host.
```
t=2017-02-01T10:24:48-0800 lvl=info msg="Migrating container" ephemeral=false statedir=/tmp/lxd_checkpoint_429974479 actionscript=true stop=true name=es-1 created=2017-01-31T19:03:54-0800
t=2017-02-01T10:24:49-0800 lvl=info msg="Failed migrating container" actionscript=true stop=true name=es-1 created=2017-01-31T19:03:54-0800 ephemeral=false statedir=/tmp/lxd_checkpoint_429974479
t=2017-02-01T10:24:49-0800 lvl=info msg="Migrated container" name=es-1 created=2017-01-31T19:03:54-0800 ephemeral=false statedir=/tmp/lxd_checkpoint_429974479 actionscript=true stop=true
```

Moving *to* this host.
```
t=2017-02-01T10:24:49-0800 lvl=info msg="Deleting container" name=es-1 creation date=2017-02-01T10:24:05-0800 ephemeral=false
t=2017-02-01T10:24:49-0800 lvl=info msg="Deleted container" name=es-1 creation date=2017-02-01T10:24:05-0800 ephemeral=false
t=2017-02-01T10:24:49-0800 lvl=info msg="Deleting container" name=es-1 creation date=2017-02-01T10:24:49-0800 ephemeral=false
t=2017-02-01T10:24:49-0800 lvl=info msg="Deleted container" name=es-1 creation date=2017-02-01T10:24:49-0800 ephemeral=false
t=2017-02-01T10:24:50-0800 lvl=info msg="Deleting container" name=es-1 creation date=2017-02-01T10:24:49-0800 ephemeral=false
t=2017-02-01T10:24:50-0800 lvl=info msg="Deleted container" name=es-1 creation date=2017-02-01T10:24:49-0800 ephemeral=false
```

Not sure why it tries to delete the failed migration 3 times, but everything seems normal considering.

**Comment by ChrisLundquist** [comment]
@stgraber @tych0 it looks like it is related to on of the LXD action_scripts / callbacks failing, the dump getting nuked, then the protocol falling over.

This is in the migration log after running `lxc move packer nucalas:`
```
(04.263599) Writing image inventory (version 1)
(04.263688) Running post-dump scripts
(04.263692) 	[/tmp/lxd_checkpoint_978884516/action.sh]
(05.664371) Error (criu/action-scripts.c:116): One of more action scripts failed
(05.664421) Post dump script passed with 1
(05.664426) Unlock network
(05.664431) Running network-unlock scripts
(05.664446) 	[/tmp/lxd_checkpoint_978884516/action.sh]
(05.674973) Unfreezing tasks into 1
(05.675000) 	Unseizing 1487 into 1
(05.675015) 	Unseizing 1717 into 1
(05.675025) 	Unseizing 1971 into 1
(05.675034) 	Unseizing 2108 into 1
(05.675085) 	Unseizing 2112 into 1
(05.675133) 	Unseizing 2113 into 1
(05.675232) 	Unseizing 2131 into 1
(05.675245) 	Unseizing 2183 into 1
(05.675254) 	Unseizing 2190 into 1
(05.675266) 	Unseizing 2196 into 1
(05.675275) 	Unseizing 2200 into 1
(05.675333) 	Unseizing 2211 into 1
(05.675345) 	Unseizing 2215 into 1
(05.675353) 	Unseizing 2217 into 1
(05.675384) 	Unseizing 2350 into 1
(05.675420) 	Unseizing 2369 into 1
(05.675431) 	Unseizing 2376 into 1
(05.675517) Writing stats
(05.675615) Dumping finished successfully
```

`lxc snapshot --stateful packer` works fine, but doesn't seem to have any callback scripts:
```
(03.986474) page-pipe: Killing page pipe
(03.986489) sk unix: Dumping external sockets
(03.986498) sk unix: 	Dumping extern: ino 0x2eb17 peer_ino 0 family    1 type    5 state  7 name (null)
(03.986519) sk unix: 	Dumped extern: id 0xcd ino 0x2eb17 peer 0 type 2 state 10 name 0 bytes
(03.986527) Writing image inventory (version 1)
(03.986809) Running post-dump scripts
(03.986816) Unlock network
(03.986820) Running network-unlock scripts
(03.993215) Unfreezing tasks into 1
(03.993236) 	Unseizing 1536 into 1
(03.993252) 	Unseizing 1656 into 1
(03.993263) 	Unseizing 1901 into 1
(03.993275) 	Unseizing 2048 into 1
(03.993286) 	Unseizing 2060 into 1
(03.993329) 	Unseizing 2068 into 1
(03.993407) 	Unseizing 2072 into 1
(03.993422) 	Unseizing 2074 into 1
(03.993433) 	Unseizing 2097 into 1
(03.993457) 	Unseizing 2099 into 1
(03.993506) 	Unseizing 2105 into 1
(03.993521) 	Unseizing 2126 into 1
(03.993532) 	Unseizing 2128 into 1
(03.993542) 	Unseizing 2277 into 1
(03.993552) 	Unseizing 2286 into 1
(03.993645) Writing stats
(03.993723) Dumping finished successfully
```

Checking what is in the failing action script is pretty hard because the folder gets cleaned up very quickly (it exists for ~2-4 seconds)

```
clundquist@nuc:~$ lxc --version
2.0.9
clundquist@nuc:~$ uname -a
Linux nuc 4.8.0-41-generic #44~16.04.1-Ubuntu SMP Fri Mar 3 17:11:16 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
clundquist@nuc:~$ dpkg -l | grep criu
ii  criu                                2.6-1ubuntu1~ubuntu16.04.2              amd64        checkpoint and restore in userspace
```

```
clundquist@NUCALASCAGE:~$ lxc --version
2.0.9
clundquist@NUCALASCAGE:~$ uname -a
Linux NUCALASCAGE 4.8.0-41-generic #44~16.04.1-Ubuntu SMP Fri Mar 3 17:11:16 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
clundquist@NUCALASCAGE:~$ dpkg -l | grep criu
ii  criu                                2.6-1ubuntu1~ubuntu16.04.2        amd64        checkpoint and restore in userspace
```

https://github.com/lxc/lxd/blob/e202b07dec34a1b93b65a2c8c8347345fc8a38b2/lxd/migrate.go#L233-L238

Appears to where the action scripts come from, I'm not sure what the parameters are when it fails, and it looks like it would take recompiling to inspect them.

**Comment by stgraber** [comment]
@brauner do you think your recent fix solves this issue? if so, we should close it.

**Comment by brauner** [comment]
@stgraber, I added a few more commits. The last one implements logging for our `netcat` implementation. I think we can close this issue for now as I think the other failing rsyncs might be related to rsync being confused about the paths. In any case, if the fixes I pushed prove not to be sufficient we can reopen.

**Comment by brauner** [comment]
For future reference, we didn't have a producer in the traditional sense but one of our users provided me with access to two LXD instances and a container that would reliably fail. I was able to track the failure down to our netcat code. @tych0 was correct in assuming that our netcat or rather golang's `io.Copy()` did not handle `EAGAIN`. However, the `EAGAIN` was not present on the unix domain socket connection but rather on stdout. Also, we shouldn't simply return a short-write on `EAGAIN` but rather keep on trying. The trick is to not retry the `io.Copy()` call which indeed is not handling this case appropriately, rather we need to handle this in the `Read()` and `Write()` calls directly.
Since I see no harm coming from it, I've wrapped all readers and writers passed to `io.Copy()` with my new `EAGAIN` wrappers for them. Let's keep an eye open for new weirdness. :)
